### PR TITLE
LF: Add tests for evaluation order on fetches

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
As for creates, these tests deliberately ignore ordering on on-catchable 
errors relative to each other since those only change error messages.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala [hunk 1]
```diff
@@ -235,13 +335,349 @@ class EvaluationOrderTest extends AnyFreeSpec with Matchers with Inside {
...
+             in fetch @M:T cId2""",
+            Array(SParty(alice)),
+            alice,
+          )
+
+          inside(res) {
+            case Success(Left(SErrorDamlException(IE.WronglyTypedContract(_, T, Dummy)))) =>
+          }
+          msgs shouldBe Seq.empty
+        }
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala:474]: I assume it is not possible to have an authorization failure for a local contract

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala [hunk 2]
```diff
@@ -235,13 +335,349 @@ class EvaluationOrderTest extends AnyFreeSpec with Matchers with Inside {
...
+            getContract = getContract,
+            getKey = getKey,
+          )
+
+          inside(res) { case Success(Left(SErrorDamlException(IE.ContractKeyNotFound(key)))) =>
+            key.templateId shouldBe T
+          }
+          msgs shouldBe globalFetchByIdTraces :+ "archive" :++ localFetchByKeyTraces
+        }
+
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala:533]: I assume it is not possible to fetch_by_key a wrongly type contract.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala [hunk 3]
```diff
@@ -166,12 +272,14 @@ class EvaluationOrderTest extends AnyFreeSpec with Matchers with Inside {
...
-        inside(res) { case SResultError(SErrorDamlException(err)) =>
-          err shouldBe a[CreateEmptyContractKeyMaintainers]
-
+        inside(res) {
+          case Success(
+                Left(SErrorDamlException(IE.CreateEmptyContractKeyMaintainers(T, _, _)))
+              ) =>
         }
-        msgs shouldBe allTraces
+        msgs shouldBe createTraces
```
Comment (cocreature) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala:280]: maybe move those inside the `inside`? It seems a bit odd to use `inside` without anything in the pattern match.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala (hunk 1 above)
Comment (cocreature) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala:466]: That’s not correct, the authorization context can change throughout a transaction. See https://gist.github.com/cocreature/e3257ce5e0855a4e0e6f7b8ea148d926 for an example.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala (hunk 2 above)
Comment (cocreature) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala:520]: I think that’s right if we assume the ledger behaves correctly but maybe still worth checking that it throws an internal error?

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala [hunk 4]
```diff
@@ -235,11 +339,291 @@ class EvaluationOrderTest extends AnyFreeSpec with Matchers with Inside {
...
+            getKey = getKey,
+          )
+          inside(res) { case Success(Left(SErrorDamlException(IE.FailedAuthorization(_, _)))) => }
+          msgs shouldBe globalFetchByKeyTraces
+        }
+      }
+
+      "a local contract" - {
+
+        // TEST_EVIDENCE: Semantics: Evaluation order of successful fetch_by_key of a local contract
```
Comment (cocreature) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala:538]: similar comment as above, it doesn’t matter if it’s local or not it matters that it’s cached. The second fetch by key of a global contract will do the same and we should test that.

**Review by cocreature — APPROVED** [review]
Nice work! A few comments

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala [hunk 5]
```diff
@@ -235,11 +339,291 @@ class EvaluationOrderTest extends AnyFreeSpec with Matchers with Inside {
...
+            e"""\(cId: ContractId M:T) -> fetch @M:T cId""",
+            Array(SContractId(cId)),
+            alice,
+            getContract = getContract,
+          )
+          inside(res) { case Success(Right(_)) => }
+          msgs shouldBe globalFetchByIdTraces
+        }
+
+        // TEST_EVIDENCE: Semantics: Evaluation order of fetch of an inactive global contract
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala:367]: I will take add test for cached/non-cached contract in separate PR.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala [hunk 6]
```diff
@@ -235,11 +339,291 @@ class EvaluationOrderTest extends AnyFreeSpec with Matchers with Inside {
...
+             in fetch @M:T cId""",
+            Array(SParty(alice), SParty(bob)),
+            alice,
+          )
+          inside(res) { case Success(Right(_)) => }
+          msgs shouldBe createTraces :++ localFetchByIdTraces
+        }
+
+        // TEST_EVIDENCE: Semantics: Evaluation order of fetch of an inactive local contract
+        "inactive contract" in {
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala:441]: I will take add test for cached/non-cached contract in separate PR.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala:547]: I will take add test for cached/non-cached contract in separate PR.

**Pull request merged**